File: prog3/trab_cpp/ArtigoPe.cpp
#include "ArtigoPe.hpp"

ArtigoPe::ArtigoPe(string natureza, string idioma, string cidade, int qtd_pags, string editora, int volume, int fasciculo, int serie, string issn): Producao("", natureza, idioma, cidade, qtd_pags, 25){
    this->editora = editora;
    this->volume = volume;
    this->fasciculo = fasciculo;
    this->serie = serie;
    this->issn = issn;
}

ArtigoPe::~ArtigoPe(){}

/* Metodo que retorna a string que contem as informacoes de uma producao em formato csv. No caso, reorna uma string no seguinte formato:
 "<Natureza>;<Idioma>;<Editora>;<Cidade>;<Volume>;<Fasciculo>;<Serie>;<ISSN>;<Paginas>"*/
string ArtigoPe::csv(){
    stringstream stream;
    stream << natureza << ";" << idioma << ";" << editora << ";" << cidade << ";" << _int_string(volume) << ";" << _int_string(fasciculo) << ";" << _int_string(serie) << ";" << issn << ";" << _int_string(qtd_pags);
    return stream.str();
}

/* Metodo que retorna a string cabecalho. */
string ArtigoPe::cabecalho(){
    return "Natureza;Idioma;Editora;Cidade;Volume;Fasciculo;Serie;ISSN;Paginas";
}

/* Operador de comparacao de uma producao com outra de mesmo subtipo. */
bool ArtigoPe::operator < (Producao& o){
    ArtigoPe& prod = (ArtigoPe&)o;
    int cmp = natureza.compare(prod.natureza);
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = idioma.compare(prod.idioma);
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = editora.compare(prod.editora);
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = cidade.compare(prod.cidade);
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = volume - prod.volume;
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = fasciculo - prod.fasciculo;
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = serie - prod.serie;
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = issn.compare(prod.issn);
    if(cmp<0) return true; else if(cmp>0) return false;
    cmp = qtd_pags - prod.qtd_pags;
    if(cmp<0) return true; else return false;

    
}

File: prog3/trab_java/utilitarios/capes/producoes/ArtigoPE.java
package utilitarios.capes.producoes;

public class ArtigoPE extends Producao implements Comparable<ArtigoPE>{
    private String editora;
    private int volume;
    private String volume_string;
    private int fasciculo;
    private String fasciculo_string;
    private int serie;
    private String serie_string;
    private String issn;
    private String pagFin;
    private String pagIni;
    

    public ArtigoPE(String natureza, String idioma, String cidade, int qtdPags, String editora, int volume, int fasciculo, int serie, String issn, String pagIni, String pagFin){
        super(null, natureza, idioma, cidade, qtdPags);
        this.editora = editora;
        this.volume = volume;
        this.volume_string = formataNum(volume);
        this.fasciculo = fasciculo;
        this.fasciculo_string = formataNum(fasciculo);
        this.serie = serie;
        this.serie_string = formataNum(serie);
        this.issn = issn;
        this.tipo = 3;
        this.pagIni = pagIni;
        this.pagFin = pagFin;
    }

    @Override
    public int compareTo(ArtigoPE o) {
        int cmp;

        cmp = this.natureza.compareTo(o.natureza); if(cmp!=0) return cmp;
        cmp = this.idioma.compareTo(o.idioma);     if(cmp!=0) return cmp;
        cmp = this.editora.compareTo(o.editora);   if(cmp!=0) return cmp;
        cmp = this.cidade.compareTo(o.cidade);     if(cmp!=0) return cmp;
        cmp = this.volume - o.volume;              if(cmp!=0) return cmp;
        cmp = this.fasciculo - o.fasciculo;        if(cmp!=0) return cmp;
        cmp = this.serie - o.serie;                if(cmp!=0) return cmp;
        cmp = this.qtdPags - o.qtdPags;            if(cmp!=0) return cmp;
        cmp = this.issn.compareTo(o.issn);                    return cmp;
    }

    public static void imprimeCabecalho(){
        System.out.println("Natureza;Idioma;Editora;Cidade;Volume;Fasciulo;Serie;ISSN;Paginas");
    }

    @Override
    public String toString(){
        return (this.natureza + ";" + this.idioma + ";" + this.editora + ";" + this.cidade + ";" + this.volume_string + ";" + this.fasciculo_string + ";" + this.serie_string + ";" + this.issn + ";" + this.qtdPags_string);
    } 
}